src/Npgsql/PoolingDataSource.cs

```diff
@@ -246,8 +246,25 @@ bool CheckIdleConnector([NotNullWhen(true)] NpgsqlConnector? connector)
         return true;
     }
 
-    internal sealed override async ValueTask<NpgsqlConnector?> OpenNewConnector(
+    internal sealed override ValueTask<NpgsqlConnector?> OpenNewConnector(
```

Comment:
Is the idea to have two versions of OpenNewConnector in order to avoid creating an NpgsqlTimeout instance for when we can't create a new connector (i.e. over-capacity mode)? I get that the NpgsqlTimeout constructor has a condition and some math but it seems a bit extreme...

Am mainly checking if there's something I'm missing, otherwise it seems like the code is getting more complex but I'm not sure it's really worth it. But am also OK to merge.

Reply:
> Is the idea to have two versions of OpenNewConnector in order to avoid creating an NpgsqlTimeout instance for when we can't create a new connector (i.e. over-capacity mode)? I get that the NpgsqlTimeout constructor has a condition and some math but it seems a bit extreme...

Yeah, though I don't think I've ever seen the actual codegen difference. The main issue is accessing `DateTime.UtcNow`, though maybe it's not that bad. I'm going to take a look.

Reply:
OK. So, with this pr we manage to inline `OpenNewConnector` (that's good) but not `GetNewConnectorNumber` (which is bad and one of the main reasons for this pr). I think we can manage that with some clever use of `MethodImpl` attribute, but the added complexity doesn't seem to be worth it (even if it will make things faster, which I'm not 100% sure). I'm going to close this pr.